### PR TITLE
Add flag `gcs-only` to start ray head without raylet

### LINKED ISSUE #19745: [Feature] Add an option to start head node without raylet
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

In a large scale ray cluster, the head node's load is high because gcs server will handle a lot of events. And the raylet on the head node make things worse.

Please add an option to start head node without raylet.

### Use case

The head node don't start raylet.

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by chenk008**
## Why are these changes needed?

In large scale ray cluster, head node is busy to handle task/actor schedule. To reduce head node load,  add a flag to start ray head without raylet.

## Related issue number

Closes https://github.com/ray-project/ray/issues/19745

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by ericl** [comment]
Hmm how is this different from setting --num-cpus=0, which should reject all tasks and actors?

File: python/ray/node.py [hunk 1]
```diff
@@ -85,7 +85,8 @@ def __init__(self,
                  head=False,
                  shutdown_at_exit=True,
                  spawn_reaper=True,
-                 connect_only=False):
+                 connect_only=False,
+                 workerless=False):
```
Comment (ericl) [inline python/ray/node.py:89]:
```suggestion
                 no_workers=False):
```

**Comment by chenk008** [comment]
> Hmm how is this different from setting --num-cpus=0, which should reject all tasks and actors?

When setting `--num-cpus=0`, the raylet still will start. And the dashboard will show the head as a worker node, it makes some users confused. 

@ericl How about using `--num-cpus=0` as workerless, so we don't need to add a `workerless` flag.

**Comment by scv119** [comment]
Can we name it gcs only mode?

**Comment by ericl** [comment]
+1 for GCS only mode

File: python/ray/node.py (hunk 1 above)
Comment (chenk008) [reply python/ray/node.py]: Thanks, I have use `no_worker` instead.

**Comment by chenk008** [comment]
@ericl @scv119 I have rename the flag to `gcs-only`, please take a look.

File: python/ray/node.py [hunk 2]
```diff
@@ -390,7 +392,9 @@ def plasma_store_socket_name(self):
     @property
     def unique_id(self):
         """Get a unique identifier for this node."""
-        return f"{self.node_ip_address}:{self._plasma_store_socket_name}"
+        if hasattr(self, "_plasma_store_socket_name"):
```
Comment (jjyao) [inline python/ray/node.py:395]: We still have things like self.raylet_ip_address when no_worker is True, which I think is not correct?

**Review by jjyao — COMMENTED** [review]
Could you help me to understand the implication of starting ray head without raylet:

1. Will the head node show up in the node_table? I think it's not?
2. If I try to start a driver on the head node, what will happen in this case?
3. When gcs_only is true, some of the `Node` properties may not make sense, for example what should `Node.address_info` be in this case?
4. Do we have any code that rely on the fact that head node has a raylet?

**Comment by jovany-wang** [comment]
> 1. Will the head node show up in the node_table? I think it's not?

No

> 2. If I try to start a driver on the head node, what will happen in this case?

It's not able to connect because there is no Raylet. Some thing like start a driver on a node which is not in Ray cluster.

> 3. When gcs_only is true, some of the Node properties may not make sense, for example what should Node.address_info be in this case?
I don't quite understand this issue. Since we don't start any raylet in the head node, we don't have any info about it.

> Do we have any code that rely on the fact that head node has a raylet?
I think No. We have been running this cluster mode in several years.

**Comment by chenk008** [comment]
@ericl @scv119 @jjyao We want to start ray head without raylet, but it still contains gcs/dashboard/autoscaler, so I think `gcs-only` is not inaccurate. How about use 'no-worker' or 'no-raylet'?

**Comment by jjyao** [comment]
@jovany-wang sorry I missed your reply.
> I don't quite understand this issue. Since we don't start any raylet in the head node, we don't have any info about it.

Even if we don't start raylet, we will still create the python `Node` object right? `Node` has some property like `raylet_ip_address`, do we need to make sure they return something sensible if raylet is not started, for example None?

**Comment by jjyao** [comment]
> @ericl @scv119 @jjyao We want to start ray head without raylet, but it still contains gcs/dashboard/autoscaler, so I think `gcs-only` is not inaccurate. How about use 'no-worker' or 'no-raylet'? WDYT?

Yea, in that sense, I agree that `gcs-only` is confusing and `no-raylet` is probably better.

> When setting --num-cpus=0, the raylet still will start. And the dashboard will show the head as a worker node, it makes some users confused.

Actually after a second thought: what if we change the dashboard to hide "empty" node by default, will this solve the problem? Seems that will be a smaller change and avoid changing the more fundamental thing that each node has a raylet.

Another question: will ray client still work if head node doesn't have a raylet?

**Comment by chenk008** [comment]
@jjyao  I think `--num-cpus=0` is not clear enough and it depends on the way the scheduler works.  

I will check whether ray client still work if head node doesn't have a raylet. If it not workes any more, maybe we can start a unschedulable raylet on the head node?

**Comment by rkooo567** [comment]
I am in favor of this change, but I think we should think about well-thought design rather than just not starting raylet. For example, since dashboard agent is not available on the GCS node, all of metrics are unavailable with this deployment model. Or ray client server is probably not working correctly. 

There could be other issues since the current “assumes”  raylet is always there. I am not sure if there are more potential issues. 

I think the right direction is to summarize what regular "ray start --head" does (what processes it starts) the roles of processes and how their lifecycle is managed. And summarize what will work/not work when raylet is not started together (and what fix is needed). Otherwise, it might cause lots of unexpected bugs.

**Comment by ericl** [comment]
What if we added a way to disable scheduling on the raylet instead of deleting the raylet? I think that would be architecturally more consistent, and may be as simple as setting --num-cpus=0 under the hood.

**Review by wilsonwang371 — COMMENTED** [review]
If gcs-only specified, it can also start monitor.py rayClientServer & dashboard, right?

**Comment by jovany-wang** [comment]
> What if we added a way to disable scheduling on the raylet instead of deleting the raylet? I think that would be architecturally more consistent, and may be as simple as setting --num-cpus=0 under the hood.

I don't think so. The node table in dashboard is not consistent with real nodes.

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by DmitriGekhtman** [comment]
I like the general direction of this PR.
What's our progress on this?

**Comment by chenk008** [comment]
@DmitriGekhtman Sorry for delay reply. As discussed above, `there could be other issues since the current “assumes” raylet is always there`, some test cases are failed ,  I think we need a clearer plan to separate components in head.

**Comment by larrylian** [comment]
Vote +1.

**Comment by wilsonwang371** [comment]
> @jjyao I think `--num-cpus=0` is not clear enough and it depends on the way the scheduler works.
> 
> I will check whether ray client still work if head node doesn't have a raylet. If it not workes any more, maybe we can start a unschedulable raylet on the head node?

Or do we need to implement a proxy to redirect the request to a node with raylet?

The ray-client server code needs access to raylet socket and object store socket. I have a graph demostrating the issue here if we have a lite-head mode that I mentioned in #22744 .  In this case, instead of running a raylet, can we change Ray client server to make it act like a proxy to another node's raylet?

Any other throught? 

image

FYI: @mwtian @iycheng @larrylian @jovany-wang

**Comment by mwtian** [comment]
IIUC, I don't see feasibility of running client server with raylet from a different node, which is equivalent to running a worker with raylet from a different node. I can be wrong, but core worker probably assumes its raylet sees the same local memory.

I think supporting not running client server or dashboard on the same node as GCS should be a useful change. Each of these services should be able to register in GCS for Ray internal routing. External routing to these services have to be taken care of by Ray operator or higher level configurations.

**Comment by rkooo567** [comment]
+1 to @mwtian. Btw I really think we need a design doc, which tells what procs can be deployed independently, what relies on what, and what would be the api look like.

**Comment by chenk008** [comment]
`procs can be deployed independently as a separate pod` sounds great!

**Pull request closed without merging**